includes/embeds/class-amp-reddit-embed-handler.php

```diff
@@ -6,70 +6,115 @@
 	}
 
 	/**
-	 * Unregister embed.
+	 * Make embed AMP compatible.
+	 *
+	 * @param DOMElement $node DOM element.
 	 */
-	public function unregister_embed() {
-		wp_embed_unregister_handler( 'amp-reddit', -1 );
+	protected function sanitize_raw_embed( DOMElement $node ) {
+		$anchor_node = $this->get_anchor_node( $node );
+		if ( null === $anchor_node ) {
+			return;
+		}
+
+		$reddit_url = $anchor_node->getAttribute( 'href' );
+		if ( empty( $reddit_url ) ) {
+			return;
+		}
+
+		$attributes = [
+			'layout'   => 'responsive',
+			'width'    => 100,
+			'height'   => 100,
+			'data-url' => $anchor_node->getAttribute( 'href' ),
+		];
+
+		$amp_node = AMP_DOM_Utils::create_node(
+			Document::fromNode( $node ),
+			$this->amp_tag,
+			$attributes
+		);
+
+		// Add the raw embed as the placeholder.
+		$placeholder_node = $node->cloneNode( true );
+		$placeholder_node->setAttribute( 'placeholder', '' );
+		$amp_node->appendChild( $placeholder_node );
+
+		$this->remove_script_sibling( $node, 'embed.redditmedia.com/widgets/platform.js' );
+		$node->parentNode->replaceChild( $amp_node, $node );
 	}
 
 	/**
-	 * Embed found with matching URL callback.
+	 * Get the first child anchor element from the blockquote element.
 	 *
-	 * @param array $matches URL regex matches.
-	 * @param array $attr    Additional parameters.
-	 * @param array $url     URL.
-	 * @return string Embed.
+	 * @param DOMElement $blockquote_node Blockquote element.
+	 * @return DOMElement|null Anchor element if found, otherwise null.
 	 */
-	public function oembed( $matches, $attr, $url ) {
-		return $this->render( [ 'url' => $url ] );
+	private function get_anchor_node( DOMElement $blockquote_node ) {
+		$first_child_element = $this->get_first_child_element( $blockquote_node );
+
+		if ( null === $first_child_element ) {
+			return null;
+		}
+
+		$anchor_node = null;
+
+		if ( 'a' === $first_child_element->nodeName ) {
+			// Anchor tag was not wrapped by `wpautop`.
+			$anchor_node = $first_child_element;
+		} elseif ( 'p' === $first_child_element->nodeName ) {
+			// If the anchor tag is wrapped by `wpautop`, it should be the first child of the `p` tag.
+			$first_child_element = $this->get_first_child_element( $first_child_element );
+
+			if ( null !== $first_child_element && 'a' === $first_child_element->nodeName ) {
+				$anchor_node = $first_child_element;
+			}
+		}
+
+		return $anchor_node;
 	}
 
 	/**
-	 * Output the Reddit amp element.
+	 * Get first child element for specified element.
 	 *
-	 * @param array $args parameters used for output.
-	 * @return string Rendered content.
+	 * @param DOMElement $node Element.
+	 * @return DOMElement|null First element if found, otherwise null.
 	 */
-	public function render( $args ) {
-		$args = wp_parse_args(
-			$args,
-			[
-				'url' => false,
-			]
-		);
-
-		if ( empty( $args['url'] ) ) {
-			return '';
+	private function get_first_child_element( DOMElement $node ) {
```

Comment:
This seems useful to be in the base class as a `protected` method. It could be called `get_first_element_child()` to reflect the DOM: https://developer.mozilla.org/en-US/docs/Web/API/ParentNode/firstElementChild